### PR TITLE
Constructor verification for heap allocation and nested structs/classes.

### PR DESCRIPTION
**Pull request by NielsMommen**
- Constructors are verified and their contract is checked when a new object is allocated on the heap.
- Constructors are checked for nested struct/class allocation. I.e., when a field in a struct/class is a struct/class.
- Currently, a default constructor has to be specified explicitly for a struct/class.
- Destructors are not supported.
- The `delete` expression is only supported when a struct/class does not have nested structs/classes.
- Allocation of structs/classes on the stack is not supported.

### CODE REVIEW

File: src/verify_expr.ml [hunk 1]
```diff
@@ -918,20 +981,23 @@ module VerifyExpr(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
...
+    structmap1 |> List.map @@ fun (sn, (sloc, sbody, spad_sym, ssize)) ->
+      let tenv = ["this", PtrType (StructType sn); current_thread_name, current_thread_type] in 
+      let body = sbody |> option_map @@ fun fields ->
+        fields |> List.map @@ function
+          | fname, (floc, fgh, ft, foffset, Some finit) ->
+            let init = check_expr_t tenv finit ft in
+            fname, (floc, fgh, ft, foffset, Some init)
+          | fd -> fd 
+      in
+    sn, (sloc, body, spad_sym, ssize)
```
Comment (btj) [inline src/verify_expr.ml:997]: Fix indentation

File: src/verify_expr.ml [hunk 2]
```diff
@@ -918,20 +981,23 @@ module VerifyExpr(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
...
-              fname, (floc, fgh, ft, foffset, Some finit) ->
-                let check_expr_t tenv e tp = check_expr_t_core functypemap funcmap [] [] ("", []) [] tenv None e tp in (* TODO: package name? *)
-                let tenv = ["this", PtrType (StructType sn); current_thread_name, current_thread_type] in 
-                let init = check_expr_t tenv finit ft in
-                fname, (floc, fgh, ft, foffset, Some init)
-            | fd -> fd 
-        in
-        sn, (sloc, body, spad_sym, ssize)
-  
+  (* CXX: check defualt initializers *)
```
Comment (btj) [inline src/verify_expr.ml:984]: typo

File: src/cxx_frontend/ast_exporter/DeclSerializer.cpp [hunk 3]
```diff
@@ -3,64 +3,81 @@
...
+    auto contractBuilder = builder.initContract(anns.size());
+    size_t i(0);
+    for (auto &ann : anns) {
+      auto annBuilder = contractBuilder[i++];
+      ser->serializeAnnotationClause(annBuilder, ann);
+    }
   }
 
-  if (decl->isThisDeclarationADefinition()) {
+  // Implicit functions have a definiton, but only have a body when
```
Comment (btj) [inline src/cxx_frontend/ast_exporter/DeclSerializer.cpp:68]: Typo

**Review by btj — CHANGES_REQUESTED** [review]
Good job, Niels! I have just a few minor comments.

File: src/ast.ml [hunk 4]
```diff
@@ -771,7 +776,25 @@ and
...
       visibility
-      
+  | CxxCtor of 
+      loc *
+      string * (* name *)
+      string * (* mangled name *)
+      (type_expr * string) list * (* params *)
+      (asn * asn) option * (* pre post *)
+      bool * (* terminates *)
+      ((string * (expr * bool (* is written *)) option) list (* init list *) * (stmt list * loc (* close brace *))) option *
```
[the remark this answers is not in the document]
Comment (NielsMommen) [reply src/ast.ml:786]: - Make two separate optional components or insert a newline for readability?
- The init list contains every field that will be initialised by the constructor invocation. If a record field has no explicit initialiser, Clang will generate one. In that case, `'field_name', Some ('ctor mangled name', false)` is present in the init list. For a primitive field that has no explicit initialiser, `'field_name', None` is present. E.g.:
  ```c++
  struct Foo {
    int _f;
  
    Foo() : _f(0) {}
    Foo(int f) : _f(f) {}
  };
  
  struct Bar {
    int _i;
    int _j = 0;
    Foo _f1;
    Foo _f2;
  
    Bar() : f2(2) {}
  };
  ```
  has following init lists
  - `Foo()`: `_f, Some ('0', true)`
  - `Foo(int f)`: `_f, Some('f', true)`
  - `Bar()`: `_i, None`; `_j, Some('0', true)`; `_f1, Some('Foo()', false)`; `_f2, Some ('Foo(2)', true)`
  Hence, the difference between `None` and `Some (..., false)` is that the former refers to a default initialisation of a primitive and the latter refers to an implicit default initialisation of a record. We need the expression generated by Clang in the latter case to know the default constructor's mangled name that should be checked.

[the remark this answers is not in the document]
Comment (btj) [reply src/ast.ml:786]: - Separate components: never mind, I had not noticed it was all one `option`.
- Init list: thanks, that clears it up!

File: src/verifast1.ml [hunk 5]
```diff
@@ -4204,9 +4215,18 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
...
+      let expr_opt = 
+        expr_opt |> option_map @@ fun e ->
+          let w, et, _ = check e in
+          assert (t = et);
+          w
+      in
+      CxxNew (l, te, expr_opt), PtrType t, None
+    | CxxConstruct (l, _, te, _) ->
+      let t = check_pure_type (pn, ilist) [] Real te in 
+      e, t, None
```
[the remark this answers is not in the document]
Comment (NielsMommen) [reply src/verifast1.ml]: Added these well formed expressions that take a `type_`.

File: src/ast.ml [hunk 6]
```diff
@@ -771,7 +776,25 @@ and
...
+      bool * (* implicit *)
+      type_ (* parent type *)
+  | CxxDtor of 
+      loc *
+      string * (* name *)
```
[the remark this answers is not in the document]
Comment (NielsMommen) [reply src/ast.ml]: Removed the name.

File: src/ast.ml [hunk 7]
```diff
@@ -771,7 +776,25 @@ and
...
+      (type_expr * string) list * (* params *)
```
[the remark this answers is not in the document]
Comment (NielsMommen) [reply src/ast.ml]: Removed the parameter list.

**Pull request merged**